# Automatidata project 
**Course 4 - The Power of Statistics**

You are a data professional in a data consulting firm, called Automatidata. The current project for their newest client, the New York City Taxi & Limousine Commission (New York City TLC) is reaching its midpoint, having completed a project proposal, Python coding work, and exploratory data analysis.

Next task: to analyze the relationship between fare amount and payment type. A follow-up email from Luana includes your specific assignment: to conduct an A/B test. 

# Course 4 End-of-course project: Statistical analysis

In this activity, you will practice using statistics to analyze and interpret data. The activity covers fundamental concepts such as descriptive statistics and hypothesis testing. You will explore the data provided and conduct A/B and hypothesis testing.  
<br/>   

**The purpose** of this project is to demostrate knowledge of how to prepare, create, and analyze A/B tests. Your A/B test results should aim to find ways to generate more revenue for taxi cab drivers.

**Note:** For the purpose of this exercise, assume that the sample data comes from an experiment in which customers are randomly selected and divided into two groups: 1) customers who are required to pay with credit card, 2) customers who are required to pay with cash. Without this assumption, we cannot draw causal conclusions about how payment method affects fare amount.

**The goal** is to apply descriptive statistics and hypothesis testing in Python. The goal for this A/B test is to sample data and analyze whether there is a relationship between payment type and fare amount. For example: discover if customers who use credit cards pay higher fare amounts than customers who use cash.
  
*This activity has four parts:*

**Part 1:** Imports and data loading
* What data packages will be necessary for hypothesis testing?

**Part 2:** Conduct EDA and hypothesis testing
* How did computing descriptive statistics help you analyze your data? 

* How did you formulate your null hypothesis and alternative hypothesis? 

**Part 3:** Communicate insights with stakeholders

* What key business insight(s) emerged from your A/B test?

* What business recommendations do you propose based on your results?

# **Conduct an A/B test**


<img src="images/Pace.png" width="100" height="100" align=left>

# **PACE stages**


Throughout these project notebooks, you'll see references to the problem-solving framework PACE. The following notebook components are labeled with the respective PACE stage: Plan, Analyze, Construct, and Execute.

<img src="images/Plan.png" width="100" height="100" align=left>


## PACE: Plan 

In this stage, consider the following questions where applicable to complete your code response:
1. What is your research question for this data project? Later on, you will need to formulate the null and alternative hypotheses as the first step of your hypothesis test. Consider your research question now, at the start of this task.


    We need to investigate if customers who use a credit card pay higher fare amounts.

*Complete the following steps to perform statistical analysis of your data:* 

### Task 1. Imports and data loading

Import packages and libraries needed to compute descriptive statistics and conduct a hypothesis test.

In [1]:
import pandas as pd
import numpy as np
from scipy import stats

In [2]:
# Load dataset into dataframe
taxi_data = pd.read_csv("2017_Yellow_Taxi_Trip_Data.csv", index_col = 0)

<img src="images/Analyze.png" width="100" height="100" align=left>

<img src="images/Construct.png" width="100" height="100" align=left>

## PACE: **Analyze and Construct**

In this stage, consider the following questions where applicable to complete your code response:
1. Data professionals use descriptive statistics for Exploratory Data Analysis. How can computing descriptive statistics help you learn more about your data in this stage of your analysis?


    Exploratory data analysis (EDA): The process of investigating, organizing, and analyzing datasets and summarizing their main characteristics, often by employing data wrangling and visualization methods; 

    Trough EDA we detect missing data, understand values distribution, outliers and make better understandig if dataset is sufficient for purpuse of the current project. 
    We also get better understanding of interactions between different variables and which variables gas greater impact on our goals.

### Task 2. Data exploration

Use descriptive statistics to conduct Exploratory Data Analysis (EDA). 

**Note:** In the dataset, `payment_type` is encoded in integers:
*   1: Credit card
*   2: Cash
*   3: No charge
*   4: Dispute
*   5: Unknown



In [3]:
taxi_data.head()

VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
24870114          2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
35634249          1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
38942136          2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
30841670          2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   

           passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
24870114                 6           3.34           1                  N   
35634249                 1           1.80           1                  N   
106203690                1           1.00           1                  N   
38942136                 1           3.70           1                  N   
30841670                 1           4.37           1                  N   

           PULocationID  DOLocationID  payment_type  fare_amount  extra  \
24870114            100           231             1         13.0    0.0   
35634249            186            43             1         16.0    0.0   
106203690           262           236             1          6.5    0.0   
38942136            188            97             1         20.5    0.0   
30841670              4           112             2         16.5    0.5   

           mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
24870114       0.5        2.76           0.0                    0.3   
35634249       0.5        4.00           0.0                    0.3   
106203690      0.5        1.45           0.0                    0.3   
38942136       0.5        6.39           0.0                    0.3   
30841670       0.5        0.00           0.0                    0.3   

           total_amount  
24870114          16.56  
35634249          20.80  
106203690          8.75  
38942136          27.69  
30841670          17.80

In [4]:
taxi_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22699 entries, 24870114 to 17208911
Data columns (total 17 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   VendorID               22699 non-null  int64  
 1   tpep_pickup_datetime   22699 non-null  object 
 2   tpep_dropoff_datetime  22699 non-null  object 
 3   passenger_count        22699 non-null  int64  
 4   trip_distance          22699 non-null  float64
 5   RatecodeID             22699 non-null  int64  
 6   store_and_fwd_flag     22699 non-null  object 
 7   PULocationID           22699 non-null  int64  
 8   DOLocationID           22699 non-null  int64  
 9   payment_type           22699 non-null  int64  
 10  fare_amount            22699 non-null  float64
 11  extra                  22699 non-null  float64
 12  mta_tax                22699 non-null  float64
 13  tip_amount             22699 non-null  float64
 14  tolls_amount           22699 non-null  float64
 1

In [6]:
taxi_data.describe(include="all")

VendorID   tpep_pickup_datetime  tpep_dropoff_datetime  \
count   22699.000000                  22699                  22699   
unique           NaN                  22687                  22688   
top              NaN  07/03/2017 3:45:19 PM  10/18/2017 8:07:45 PM   
freq             NaN                      2                      2   
mean        1.556236                    NaN                    NaN   
std         0.496838                    NaN                    NaN   
min         1.000000                    NaN                    NaN   
25%         1.000000                    NaN                    NaN   
50%         2.000000                    NaN                    NaN   
75%         2.000000                    NaN                    NaN   
max         2.000000                    NaN                    NaN   

        passenger_count  trip_distance    RatecodeID store_and_fwd_flag  \
count      22699.000000   22699.000000  22699.000000              22699   
unique              NaN            NaN           NaN                  2   
top                 NaN            NaN           NaN                  N   
freq                NaN            NaN           NaN              22600   
mean           1.642319       2.913313      1.043394                NaN   
std            1.285231       3.653171      0.708391                NaN   
min            0.000000       0.000000      1.000000                NaN   
25%            1.000000       0.990000      1.000000                NaN   
50%            1.000000       1.610000      1.000000                NaN   
75%            2.000000       3.060000      1.000000                NaN   
max            6.000000      33.960000     99.000000                NaN   

        PULocationID  DOLocationID  payment_type   fare_amount         extra  \
count   22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean      162.412353    161.527997      1.336887     13.026629      0.333275   
std        66.633373     70.139691      0.496211     13.243791      0.463097   
min         1.000000      1.000000      1.000000   -120.000000     -1.000000   
25%       114.000000    112.000000      1.000000      6.500000      0.000000   
50%       162.000000    162.000000      1.000000      9.500000      0.000000   
75%       233.000000    233.000000      2.000000     14.500000      0.500000   
max       265.000000    265.000000      4.000000    999.990000      4.500000   

             mta_tax    tip_amount  tolls_amount  improvement_surcharge  \
count   22699.000000  22699.000000  22699.000000           22699.000000   
unique           NaN           NaN           NaN                    NaN   
top              NaN           NaN           NaN                    NaN   
freq             NaN           NaN           NaN                    NaN   
mean        0.497445      1.835781      0.312542               0.299551   
std         0.039465      2.800626      1.399212               0.015673   
min        -0.500000      0.000000      0.000000              -0.300000   
25%         0.500000      0.000000      0.000000               0.300000   
50%         0.500000      1.350000      0.000000               0.300000   
75%         0.500000      2.450000      0.000000               0.300000   
max         0.500000    200.000000     19.100000               0.300000   

        total_amount  
count   22699.000000  
unique           NaN  
top              NaN  
freq             NaN  
mean       16.310502  
std        16.097295  
min      -120.300000  
25%         8.750000  
50%        11.800000  
75%        17.800000  
max      1200.290000

You are interested in the relationship between payment type and the fare amount the customer pays. One approach is to look at the average fare amount for each payment type. 

In [23]:
a = taxi_data.groupby("payment_type")["fare_amount"].mean()
a.iloc[0]

13.429747789059942

In [26]:
print(f"paid with credit cards, average amount of fare: {a.iloc[0]:.2f} USD")
print(f"paid with cash, average amount of fare: {a.iloc[1]:.2f} USD")

paid with credit cards, average amount of fare: 13.43 USD
paid with cash, average amount of fare: 12.21 USD


Based on the averages shown, it appears that customers who pay in credit card tend to pay a larger fare amount than customers who pay in cash. However, this difference might arise from random sampling, rather than being a true difference in fare amount. To assess whether the difference is statistically significant, you conduct a hypothesis test.


### Task 3. Hypothesis testing

Before you conduct your hypothesis test, consider the following questions where applicable to complete your code response:

1. Recall the difference between the null hypothesis and the alternative hypotheses. Consider your hypotheses for this project as listed below.

$H_0$: There is no difference in the average fare amount between customers who use credit cards and customers who use cash.

$H_A$: There is a difference in the average fare amount between customers who use credit cards and customers who use cash.



Your goal in this step is to conduct a two-sample t-test. Recall the steps for conducting a hypothesis test: 


1.   State the null hypothesis and the alternative hypothesis
2.   Choose a signficance level
3.   Find the p-value
4.   Reject or fail to reject the null hypothesis 



**Note:** For the purpose of this exercise, your hypothesis test is the main component of your A/B test. 

You choose 5% as the significance level and proceed with a two-sample t-test.

In [8]:
taxi_data[taxi_data["payment_type"] == 1]

VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
24870114          2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
35634249          1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
38942136          2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
23345809          2   03/25/2017 8:34:11 PM   03/25/2017 8:42:11 PM   
...             ...                     ...                     ...   
60425673          1   07/16/2017 3:22:51 AM   07/16/2017 3:40:52 AM   
67858616          2  08/10/2017 10:20:04 PM  08/10/2017 10:29:31 PM   
66632549          2   08/06/2017 4:43:59 PM   08/06/2017 5:24:47 PM   
60217333          2  07/15/2017 12:56:30 PM   07/15/2017 1:08:26 PM   
17208911          1   03/02/2017 1:02:49 PM   03/02/2017 1:16:09 PM   

           passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
24870114                 6           3.34           1                  N   
35634249                 1           1.80           1                  N   
106203690                1           1.00           1                  N   
38942136                 1           3.70           1                  N   
23345809                 6           2.30           1                  N   
...                    ...            ...         ...                ...   
60425673                 1           5.70           1                  N   
67858616                 1           0.89           1                  N   
66632549                 1          16.71           2                  N   
60217333                 1           2.36           1                  N   
17208911                 1           2.10           1                  N   

           PULocationID  DOLocationID  payment_type  fare_amount  extra  \
24870114            100           231             1         13.0    0.0   
35634249            186            43             1         16.0    0.0   
106203690           262           236             1          6.5    0.0   
38942136            188            97             1         20.5    0.0   
23345809            161           236             1          9.0    0.5   
...                 ...           ...           ...          ...    ...   
60425673            249            17             1         19.0    0.5   
67858616            229           170             1          7.5    0.5   
66632549            132           164             1         52.0    0.0   
60217333             68           144             1         10.5    0.0   
17208911            239           236             1         11.0    0.0   

           mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
24870114       0.5        2.76          0.00                    0.3   
35634249       0.5        4.00          0.00                    0.3   
106203690      0.5        1.45          0.00                    0.3   
38942136       0.5        6.39          0.00                    0.3   
23345809       0.5        2.06          0.00                    0.3   
...            ...         ...           ...                    ...   
60425673       0.5        4.05          0.00                    0.3   
67858616       0.5        1.76          0.00                    0.3   
66632549       0.5       14.64          5.76                    0.3   
60217333       0.5        1.70          0.00                    0.3   
17208911       0.5        2.35          0.00                    0.3   

           total_amount  
24870114          16.56  
35634249          20.80  
106203690          8.75  
38942136          27.69  
23345809          12.36  
...                 ...  
60425673          24.35  
67858616          10.56  
66632549          73.20  
60217333          13.00  
17208911          14.15  

[15265 rows x 17 columns]

    H_0: There is no difference in the average fare amount between customers who use credit cards and customers who use cash.   
    H_A: There is a difference in the average fare amount between customers who use credit cards and customers who use cash.

In [9]:
# sample 1 - users who pay by credit card
taxi_data_credit = taxi_data[taxi_data["payment_type"] == 1]
# sample 2 - users who pay by credit card
taxi_data_cash = taxi_data[taxi_data["payment_type"] == 2]

In [15]:
# set significant level to 5%
significant_level = 0.05
# Calculate the p-value
stats.ttest_ind(a=taxi_data_credit["fare_amount"],
                 b=taxi_data_cash["fare_amount"],
                 equal_var=False,
                 alternative="two-sided")

Ttest_indResult(statistic=6.866800855655372, pvalue=6.797387473030518e-12)

In [13]:
# Calculate the p-value
stats.ttest_ind(a=taxi_data_credit["fare_amount"],
                 b=taxi_data_cash["fare_amount"],
                 equal_var=False,
                 alternative="greater")

Ttest_indResult(statistic=6.866800855655372, pvalue=3.398693736515259e-12)

==> ENTER YOUR DECISION TO ACCEPT OR REJECT THE NULL HYPOTHESIS

    As p-value = 6.8 * 10^-12 which is much less than our significant_level equal 0.05 we can reject our null hypothesis    
    so, we conduct that There is a difference in the average fare amount between customers who use credit cards and customers who use cash.

You conclude that there is a statistically significant difference in the average fare amount between customers who use credit cards and customers who use cash.

<img src="images/Execute.png" width="100" height="100" align=left>

## PACE: **Execute**

Consider the questions in your PACE Strategy Document to reflect on the Execute stage.

### Task 4. Communicate insights with stakeholders

*Ask yourself the following questions:*

1. What business insight(s) can you draw from the result of your hypothesis test?

        According A/B test we conduct there is differences between users who paid by credit card and by cash
        It can be reccommend to advertise payment by credit card

2. Consider why this A/B test project might not be realistic, and what assumptions had to be made for this educational project.

        The conducted A/B test can be not realistic as we made assumption that the sample data comes from an experiment in      
        which customers are randomly selected and divided into two groups:      
                1) customers who are required to pay with credit card,         
                2) customers who are required to pay with cash. Without this assumption, we cannot draw causal
                conclusions about how payment method affects fare amount.
        So in reality we need to design our experiment properly
